apps/leo_gateway/src/leo_nfs_readdir_state_ets.erl

```diff
@@ -0,0 +1,130 @@
+-include_lib("leo_commons/include/leo_commons.hrl").
+-include_lib("stdlib/include/ms_transform.hrl").
+
+%% gen_server callbacks
+-export([init/1,
+         handle_call/3,
+         handle_cast/2,
+         handle_info/2,
+         terminate/2,
+         code_change/3]).
+-export([info/0, start_link/1]).
+-export([add_readdir_entry/2, get_readdir_entry/1, del_readdir_entry/1]).
+
+-record(state, {
+          scan_interval :: integer(),
+          entry_ttl     :: integer()
+         }).
+
+
+-define(LEO_GW_NFS_READDIR_ENTRY_ETS_TBL, readdir_entry_ets_tbl).
+
+info() ->
+    ets:info(?LEO_GW_NFS_READDIR_ENTRY_ETS_TBL).
+
+
+start_link(Params) ->
+    ets:new(?LEO_GW_NFS_READDIR_ENTRY_ETS_TBL, [set, named_table, public]),
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [Params], []).
+
+
+%% @doc
+add_readdir_entry(CookieVerf, ReadDirEntry) ->
+    Now = leo_date:now(),
+    gen_server:call(?MODULE, scan),
+    ets:insert(?LEO_GW_NFS_READDIR_ENTRY_ETS_TBL,
+               {CookieVerf, {Now, ReadDirEntry}}).
+
+
+%% @doc
+get_readdir_entry(CookieVerf) ->
+    Ret = ets:lookup(?LEO_GW_NFS_READDIR_ENTRY_ETS_TBL, CookieVerf),
+    case Ret of
+        [] ->
+            not_found;
+        [{_, {_, ReadDirEntry}}|_] ->
+            Now = leo_date:now(),
+            ets:insert(?LEO_GW_NFS_READDIR_ENTRY_ETS_TBL, {CookieVerf, {Now, ReadDirEntry}}),
+            {ok, ReadDirEntry}
+    end.
+
+
+%% @doc
+del_readdir_entry(CookieVerf) ->
+    ets:delete(?LEO_GW_NFS_READDIR_ENTRY_ETS_TBL, CookieVerf).
+
+%% ---------------------------------------------------------------------
+%% GEN_SERVER CALLBACKS
+%% ---------------------------------------------------------------------
+init([Args]) ->
+    ScanInterval = leo_misc:get_value('nfsd_readdir_scan_int', Args, ?DEF_NFSD_READDIR_SCAN_INT) * 1000,
+    EntryTTL = leo_misc:get_value('nfsd_readdir_entry_ttl', Args, ?DEF_NFSD_READDIR_ENTRY_TTL),
+    ?debug("init/1", "Scan Int: ~p, TTL: ~p", [ScanInterval, EntryTTL]),
+    erlang:send_after(ScanInterval, self(), scan),
+    {ok, #state{scan_interval = ScanInterval,
+                entry_ttl = EntryTTL}}.
+
+
+handle_call(scan, _, #state{entry_ttl = EntryTTL} = S) ->
+    cleanup(EntryTTL),
+    {reply, ok, S}.
+
+
+handle_cast(_, S) ->
+    {noreply, S}.
+
+
+handle_info(scan, #state{scan_interval = ScanInterval,
+                         entry_ttl = EntryTTL} = S) ->
+    cleanup(EntryTTL),
+    erlang:send_after(ScanInterval, self(), scan),
+    {noreply, S}.
+
+terminate(_Reason, _S) ->
+    ok.
+
+
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+
+
+%% ---------------------------------------------------------------------
+%% PRIVATE METHODS 
+%% ---------------------------------------------------------------------
+cleanup(EntryTTL) ->
+    Now = leo_date:now(),
+    Limit = Now - EntryTTL,
+    MatchSpec = ets:fun2ms(fun({Cookie, {Time, _}}) when Time < Limit ->
+                                   true
+                           end),
```

Comment:
Kind of a tiny improvement but as it's not negligible to compile a function into a match spec,
Calculating a match spec at init and storing it into State of this gen_server then just referencing here would be efficient.
Need to change to
```erlang
fun({_Cookie, {Time, _}}) ->
    Limit = leo_date:now() - EntryTTL,
    Time < Limit
end
```
from
```erlang
fun({_Cookie, {Time, _}}) when Time < Limit ->
    true
end
```